### PR TITLE
cmd, core, pm: O validates B's expected price for a payment

### LINKED ISSUE #955: O should validate pricePerPixel sent by B
**Is your feature request related to a problem? Please describe.**

(To be filled in)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by kyriediculous**
**What does this pull request do? Explain your changes. (required)**
This PR allows an orchestrator to check whether a broadcaster's `payment.ExpectedPrice` matches the dynamically generated `PriceInfo` broadcaster based on the orchestrator's `basePrice` and transaction cost overhead. 

If the `payment.ExpectedPrice` is outdated (doesn't match) this will count as an acceptable error , similar to acceptable errors for pm tickets, as long as the broadcaster is still within the grace period (hasn't reached the maximum error count). 

The orchestrator uses a shared  error counter defined on the `LivepeerNode`  for both acceptable price errors and acceptable pm ticket errors. 

**Specific updates (required)**
- Created (extracted out of `SenderMonitor`)  an `ErrorMonitor` that tracks both acceptable pm ticket errors as well as acceptable price errors.
- Added a price check in `orch.ProcessPayment`
- Added an unexported helper method on `orchestrator` to check expected price vs the orchestrator's price for a sender
- Added an `ErrorMonitor` interface in the `pm` package
- Added a `stubErrorMonitor` to `pm` stubs/mocks

**How did you test each of these updates (required)**
Added unit tests

**Does this pull request close any open issues?**
Fixes #955 

**Checklist:**

- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: core/errormonitor.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+type errorMonitor struct {
+	mu             sync.RWMutex
+	maxErrCount    int
+	errCount       map[ethcommon.Address]int
+	gasPriceUpdate chan struct{}
+}
+
+// NewErrorMonitor returns a enw errorMonitor instance
```
Comment (yondonfu) [inline core/errormonitor.go]: "enw" -> "new"

File: core/errormonitor.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+// ClearErrCount zeroes the error count for a sender
+func (em *errorMonitor) ClearErrCount(sender ethcommon.Address) {
+	em.mu.Lock()
+	defer em.mu.Unlock()
+	em.errCount[sender] = 0
+}
+
+// ResetErrCounts clears error counts for all senders
+func (em *errorMonitor) ResetErrCounts() {
```
Comment (yondonfu) [inline core/errormonitor.go]: I don't think this needs to be exported

File: core/errormonitor.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
+package core
+
+import (
+	"sync"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+type errorMonitor struct {
+	mu             sync.RWMutex
```
Comment (yondonfu) [inline core/errormonitor.go]: Doesn't look we exclusively read lock/unlock anywhere so I think this can be `sync.Mutex` instead

File: core/errormonitor.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// Start creates a new worker to watch for gasPrice updates
+func (em *errorMonitor) Start() {
+	go em.startGasPriceUpdateLoop()
+}
+
+// Stop clears all errors and stops the gasPrice update watcher
+func (em *errorMonitor) Stop() {
+	close(em.gasPriceUpdate)
```
Comment (yondonfu) [inline core/errormonitor.go]: Currently, the `gasPriceUpdate` channel is actually created by `eth.GasPriceMonitor` and then passed into `errorMonitor` so if `errorMonitor` closes the channel and for some reason `eth.GasPriceMonitor` is still running a send into the closed `gasPriceUpdate` channel would panic. Furthermore, generally, the sender of a channel should be the one to close the channel to signal to a receiver that there are no more values to receive. So, I don't think `errorMonitor` should close the channel and instead we should update `eth.GasPriceMonitor` to close the channel when it is stopped.

File: core/errormonitor.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	defer em.mu.Unlock()
+	// Init a fresh map
+	em.errCount = make(map[ethcommon.Address]int)
+}
+
+// startGasPriceUpdateLoop initiates a loop that runs a worker
+// to reset the errCount for senders every time a gas price change
+// notification is received
+func (em *errorMonitor) startGasPriceUpdateLoop() {
+	defer em.ResetErrCounts()
```
Comment (yondonfu) [inline core/errormonitor.go]: Why do we need this deferred `em.ResetErrCounts()`? IIUC, this call would reset the error counts when `em.gasPriceUpdate` is closed, but we only need to reset the error counts when we receive an actual update from the channel. If `em.gasPriceUpdate` is closed because `em.GasPriceMonitor` is stopped (likely because the program is exiting), then any remaining values in `em.errCount` won't be retained anyway.

File: core/orchestrator.go [hunk 6]
```diff
@@ -222,6 +234,23 @@ func (orch *orchestrator) DebitFees(manifestID ManifestID, price *net.PriceInfo,
 	orch.node.Balances.Debit(manifestID, priceRat.Mul(priceRat, big.NewRat(pixels, 1)))
 }
 
+// Acceptable price checks whether the payment sender's expected price sent with a payment is acceptable
+func (orch *orchestrator) acceptablePrice(sender ethcommon.Address, ep *net.PriceInfo) (bool, error) {
+	epRat := big.NewRat(ep.GetPricePerUnit(), ep.GetPixelsPerUnit())
```
Comment (yondonfu) [inline core/orchestrator.go:257]: Should check that `ep.GetPixelsPerUnit() > 0` since `ep` is supplied by B and it could be the case that `ep = nil` or `ep.PixelsPerUnit = 0`

**Review by yondonfu — CHANGES_REQUESTED** [review]
Left a few comments as a part of a first pass review. Before taking a second pass I wanted to highlight the following:

Currently, this PR implements the acceptable price check in `ProcessPayment()` - if the error is acceptable we proceed and process the tickets in the payment, else we return an error from `ProcessPayment()` and do not process the tickets in the payment. The problem is that even if `ProcessPayment()` returns an error, at the moment O will continue execution of `ServeSegment()`. An initial thought might be that this is okay because if an invalid price triggers a non-acceptable error then `ProcessPayment()` would not credit B and eventually B's balance will become insufficient preventing any further transcoding from taking place. However, the following attack is still possible:

1. B sends a payment worth EV with a valid price
2. B sends a payment worth 0 (no tickets) with price = 0
3. Repeat step 2 until B exceeds the max error count
4. B sends a payment worth 0 (no tickets) with price = 0

After step 1, at no point does B's balance drop below EV because starting at step 2, B sends a payment with a price = 0 which means no fees are ever debited from B's balance. In step 4, `ProcessPayment()` will return an error since B has exceeded the max error count. However, O will still transcode B's video because B's balance is still sufficient at this point because it never dropped below EV after step 1.

If B exceeds the max error count and the price it includes in the payment is invalid, O should not download/transcode the segment. A few initial thoughts on how to do this:

1. Add an additional `Orchestrator` method for checking the price and if an error is returned, `ServeSegment()` will halt execution
2. Rename `SufficientBalance()` to something like `ValidatePriceAndBalance()` and update the method to not only check for a sufficient balance, but also to check for a valid price. If this method returns an error then `ServeSegment()` will halt execution
3. Add a way to signal from `ProcessPayment()` whether to halt execution or not. This could be accomplished by adding an additional boolean return value or by wrapping the error returned to include a boolean (checking it would be similar to what we do with the `pm.Error` type assertion in `ProcessPayment())`.

My preferred approach of the above options is 2, but perhaps there are better approaches.

**Comment by kyriediculous** [comment]
I favor pulling all acceptable error related types and logic into their own package

**errors:** 
* acceptable.go
* errormonitor.go

We can then check whether the error returned from `ProcessPayment` is actually acceptable because at that point it doesn't matter whether is a receiveError or an expectedPriceError, just whether it is acceptable.

We can also use the error package in core, pm and server without facing cyclic imports. 

Quick prototype: https://github.com/livepeer/go-livepeer/commit/a09aa8387e7cdec0aaccfbe75c3ad2a0150ecaa1

**Comment by j0sh** [comment]
> I favor pulling all acceptable error related types and logic into their own package

Not in favor of this. Not only do we probably have too many packages already - I'd actually prefer fewer, since then it's that much easier to settle the increasingly common question of "where to put this new helper function?". Moreover, having a "PM helper that's not a Livepeer helper, but still used by Livepeer"  really sets an undesirable precedent in fracturing package responsibilities. Some thoughts when we had an earlier discussion about this:

https://github.com/livepeer/go-livepeer/pull/702#discussion_r250301294

In general, I'd prefer a bit of repetition over bending bendwards to stay DRY and accommodate the golang cyclic checker. Or bite the bullet, unwind the strained inversion of control, and have `pm` import `common`.

**Comment by kyriediculous** [comment]
> "PM helper that's not a Livepeer helper, but still used by Livepeer"

Acceptable errors are not specific to the `pm` package anymore though, acceptable ticket receive errors are, but pricing errors can also be acceptable which is pertained to the accounting logic.

While I originally intended for this to go into the `common` package rather than its own, this isn't possible as it imports the `pm` package solely for the ticket struct for db storage. Then again, the SQL specific database logic shouldn't be in the `common` package as a whole either, in my opinion, now that we are going down the package structure rabit hole 🐰. (edit: you mention this further in the discussion as well, my bad didn't get to that far yet when writing the reply)

The original idea of making the `pm` package extractable is a noble one. It is however a client implementation and different projects might have different client-side requirements for the pm protocol, I'm not sure about this. As I understood from discussions with Yondon about this, I thought we had abandoned that route for now. However I would like to keep it as extractable as possible so how about an alternative: we could place `AcceptableErrors` in `core` and keep `ReceiveErrors` in the `pm` package so we do not create that dependency. But since they adhere to the same interface we can still check if a receiveError is acceptable. This is similar to what @yondonfu described as proposal three, albeit a little more formal code-wise. In my opinion the other two proposed solutions circumvent a larger problem at hand, but require less work.

File: core/orchestrator.go [hunk 7]
```diff
@@ -236,7 +236,7 @@ func (orch *orchestrator) DebitFees(manifestID ManifestID, price *net.PriceInfo,
 
 // Acceptable price checks whether the payment sender's expected price sent with a payment is acceptable
 func (orch *orchestrator) acceptablePrice(sender ethcommon.Address, ep *net.PriceInfo) (bool, error) {
-	if ep == nil || ep.GetPixelsPerUnit() < 0 {
+	if ep == nil || ep.GetPixelsPerUnit() <= 0 {
```
Comment (yondonfu) [inline core/orchestrator.go:254]: Since we're checking that `ep.GetPixelsPerUnit() <= 0` I think we should also have a test case for when the value is negative (looks like we only have one for when the value is 0 right now)

File: eth/gaspricemonitor.go [hunk 8]
```diff
@@ -112,6 +112,9 @@ func (gpm *GasPriceMonitor) Stop() error {
 	gpm.cancel = nil
 	gpm.polling = false
 
+	// Close the update channel when gpm is stopped
+	close(gpm.update)
```
Comment (yondonfu) [inline eth/gaspricemonitor.go:116]: Can we add a test that checks for this close?

**Comment by yondonfu** [comment]
I don't think we need to export any error struct types.

> we could place AcceptableErrors in core and keep ReceiveErrors in the pm package so we do not create that dependency. But since they adhere to the same interface we can still check if a receiveError is acceptable.

Yeah we could remove the `pm.Error` interface and define a `core.AcceptableError` interface and have acceptable errors returned by both the `pm` package and the `core` package adhere to this interface.

File: eth/gaspricemonitor.go (hunk 8 above)
Comment (kyriediculous) [reply eth/gaspricemonitor.go:116]: added in https://github.com/livepeer/go-livepeer/pull/991/commits/b9d2fae0b21dd245bd209c43f55a989f8dd36f34

File: core/orchestrator.go (hunk 7 above)
Comment (kyriediculous) [reply core/orchestrator.go:254]: added in https://github.com/livepeer/go-livepeer/pull/991/commits/b9d2fae0b21dd245bd209c43f55a989f8dd36f34

**Comment by kyriediculous** [comment]
So the error handling has been addressed in a few subsequent commits.

In the rebase process I would possibly like to squash following commits but keep the squashed version as a seperate commit in the history:

```
pm, core: server: AcceptableError interface
server, core: process payment acceptable errors
```

File: core/orchestrator.go [hunk 9]
```diff
@@ -109,13 +109,21 @@ func (orch *orchestrator) ProcessPayment(payment net.Payment, manifestID Manifes
...
+		acceptablePrice        bool
+		unacceptableReceiveErr bool
+		didReceiveErr          bool
 	)
 
-	if okPrice, acceptablePriceErr = orch.acceptablePrice(ethcommon.BytesToAddress(payment.Sender), payment.GetExpectedPrice()); !okPrice {
-		return acceptablePriceErr
+	err := orch.acceptablePrice(ethcommon.BytesToAddress(payment.Sender), payment.GetExpectedPrice())
+	if err != nil {
+		glog.Errorf("%s", err.Error())
```
Comment (yondonfu) [inline core/orchestrator.go]: Can shorten to `glog.Error(err)`

File: core/orchestrator.go [hunk 10]
```diff
@@ -109,13 +109,21 @@ func (orch *orchestrator) ProcessPayment(payment net.Payment, manifestID Manifes
...
-		didReceiveErr      bool
+		didPriceErr            bool
+		acceptablePrice        bool
+		unacceptableReceiveErr bool
+		didReceiveErr          bool
 	)
 
-	if okPrice, acceptablePriceErr = orch.acceptablePrice(ethcommon.BytesToAddress(payment.Sender), payment.GetExpectedPrice()); !okPrice {
-		return acceptablePriceErr
+	err := orch.acceptablePrice(ethcommon.BytesToAddress(payment.Sender), payment.GetExpectedPrice())
```
Comment (yondonfu) [inline core/orchestrator.go:117]: Can use `if err := ... ; err != nil {}` syntax to make this more concise

File: pm/stub.go [hunk 11]
```diff
@@ -444,6 +444,13 @@ type MockReceiveError struct {
 	acceptable bool
 }
 
+type mockReceiveError interface {
```
Comment (yondonfu) [inline pm/stub.go]: Since we usually mock/stub the struct that implements an interface (as opposed to creating a mock/stub interface) and since we already have a type called `MockReceiveError`, could we name this something else? Maybe just `acceptableError`?

File: server/segment_rpc.go [hunk 12]
```diff
@@ -59,24 +59,30 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
...
-	// to the broadcaster
+	// oInfo will be non-nil if we need to send an updated net.OrchestratorInfo to the broadcaster
 	var oInfo *net.OrchestratorInfo
 
-	if err := orch.ProcessPayment(payment, segData.ManifestID); err != nil {
-		glog.Errorf("Error processing payment: %v", err)
+	if paymentError := orch.ProcessPayment(payment, segData.ManifestID); paymentError != nil {
 
+		acceptableErr, ok := paymentError.(core.AcceptableError)
+		if !ok || !acceptableErr.Acceptable() {
```
Comment (yondonfu) [inline server/segment_rpc.go]: Noting that this will result in a HTTP error when `payment.Sender = nil || len(payment.Sender) == 0` and when `payment.TicketParams = nil` which would be the case for empty payment headers. While we don't send empty payments right now (since B usually doesn't credit its balance past the minimum), it might be useful to allow empty payments as long as the balance is sufficient and the price is acceptable (small change to `if ok && !acceptableErr.Acceptable()`)

File: server/segment_rpc.go [hunk 13]
```diff
@@ -59,24 +59,30 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
 		oInfo, err = orchestratorInfo(orch, getPaymentSender(payment), orch.ServiceURI().String())
 		if err != nil {
 			glog.Errorf("Error updating orchestrator info: %v", err)
 			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
 			return
 		}
+
+		glog.Errorf("Acceptable error occured when processing payment: %v", paymentError.Error())
```
Comment (yondonfu) [inline server/segment_rpc.go]: Can pass in `paymentError` instead of `paymentError.Error()`

File: server/segment_rpc.go [hunk 14]
```diff
@@ -59,24 +59,30 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
...
 	var oInfo *net.OrchestratorInfo
 
-	if err := orch.ProcessPayment(payment, segData.ManifestID); err != nil {
-		glog.Errorf("Error processing payment: %v", err)
+	if paymentError := orch.ProcessPayment(payment, segData.ManifestID); paymentError != nil {
 
+		acceptableErr, ok := paymentError.(core.AcceptableError)
+		if !ok || !acceptableErr.Acceptable() {
+			glog.Errorf("Unacceptable error occured processing payment: %v", paymentError.Error())
+			http.Error(w, fmt.Sprintf("Internal Server Error: %s", paymentError.Error()), http.StatusInternalServerError)
```
Comment (yondonfu) [inline server/segment_rpc.go]: `http.StatusBadRequest` is properly a better fit here (if we make the switch we can also remove the "Internal Server Error" prefix)

File: server/segment_rpc.go [hunk 15]
```diff
@@ -59,24 +59,30 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
...
 			return
 		}
+
+		glog.Errorf("Acceptable error occured when processing payment: %v", paymentError.Error())
 	}
 
 	if !orch.SufficientBalance(segData.ManifestID) {
 		glog.Errorf("Insufficient credit balance for stream with manifestID %v\n", segData.ManifestID)
-		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		http.Error(w, "Internal Server Error: insufficient balance", http.StatusInternalServerError)
```
Comment (yondonfu) [inline server/segment_rpc.go]: `http.StatusBadRequest` is properly a better fit here (if we make the switch we can also remove the "Internal Server Error" prefix)

File: core/orchestrator.go (hunk 10 above)
Comment (kyriediculous) [reply core/orchestrator.go:117]: I think that makes `err` scoped to that `if` block if I'm not mistaken and we need to access it at line 124, outside of that particular if block.

File: server/segment_rpc.go (hunk 15 above)
Comment (kyriediculous) [reply server/segment_rpc.go]: Fixed in aceab05

File: server/segment_rpc.go (hunk 14 above)
Comment (kyriediculous) [reply server/segment_rpc.go]: Fixed in aceab05

File: server/segment_rpc.go (hunk 13 above)
Comment (kyriediculous) [reply server/segment_rpc.go]: Fixed in aceab05

File: server/segment_rpc.go (hunk 12 above)
Comment (kyriediculous) [reply server/segment_rpc.go]: Fixed in aceab05

File: pm/stub.go (hunk 11 above)
Comment (kyriediculous) [reply pm/stub.go]: Fixed in aceab05

File: core/orchestrator.go (hunk 9 above)
Comment (kyriediculous) [reply core/orchestrator.go]: Fixed in aceab05

File: core/orchestrator.go (hunk 10 above)
Comment (yondonfu) [reply core/orchestrator.go:117]: Ah yes you're right - forgot that we need `err` for the type assertion on L124

**Comment by yondonfu** [comment]
Needs a rebase!

**Comment by kyriediculous** [comment]
Rebased, are you okay with keeping the 4 seperate commits?

**Review by yondonfu — APPROVED** [review]
> are you okay with keeping the 4 seperate commits?

👍 

LGTM :shipit:
